Improve error handling, cleanup

sourcejedi opened the pull request:
@cgwalters suggested the cleanup attribute could be used to make sure we
don't leak resources when prepare_next_transfer() fails, as this was overlooked in
the past.
#77 (comment)

It took a while to untangle to my satisfaction, but I think it helps.  It turns out we'd still overlooked closing the one fd in there :).

At the same I noticed a "few" other error paths which could be improved.  Enjoy!

Tojaj commented on librepo/downloader.c:
```diff
@@ -1329,7 +1330,7 @@ check_finished_transfer_checksum(int fd,
                                               &calculated,
                                               err);
         if (!ret)
```
IMHO, this won't work properly.

The ``ret`` which is tested here is a ``ret`` var defined in this context (context of the ``for`` loop during call of ``lr_checksum_fd_compare``). The newly defined ``ret`` in the main context of ``check_finished_transfer_checksum`` function will still be ``TRUE`` (it won't be set to FALSE) and after the control jump to the ``cleanup`` it will return ``TRUE`` and ``GError`` instead of ``FALSE`` and ``GError``.

sourcejedi replied:
Indeed.  This should finally be fixed now, as per your instructions.

Tojaj commented on librepo/downloader.c:
```diff
@@ -1329,7 +1330,7 @@ check_finished_transfer_checksum(int fd,
                                               &calculated,
                                               err);
         if (!ret)
```
To fix this, just replace:

    gboolean ret = lr_checksum_fd_compare(chksum->type,

with:

    ret = lr_checksum_fd_compare(chksum->type,

ignatenkobrain commented:
ping @sourcejedi ?

sourcejedi commented:
Sorry for the wait after you'd reviewed this. I could see you were right about the usage of `ret`. IIRC, I made a mistake when rebasing that change.

As a result, I wanted to check things over again.  (Which is a bit of a job, and I never got round to it).  I should be able to get it sorted this weekend.

So far I've read over all the other changes to `downloader.c`, and I'm happy with them.  (Also I've noticed it would be warned about by `CFLAGS=-Wshadow`, but this is not included in `-Wall`. Apparently -Wshadow is not included in -Wextra either.  -Wshadow also detects one other, harmless "shadowed" variable).

sourcejedi commented:
Done!

ignatenkobrain commented:
Jenkins, retest this please

Conan-Kudo commented:
Does any of this still apply anymore? Or should this just be closed?

sourcejedi commented:
Sure, everything except downloader.c applies.  The addition of `#ifdef WITH_ZCHUNK` in prepare_next_transfer() felt a bit unfortunate - adding yet another code path in there :-), that does `return TRUE`.  But I think we just keep that specific path how it is, and the merge conflict will resolve fine.

I might do this twice & link you a history that does a merge instead of a rebase. I think my tools are better at merging, but I assume you'll want a full rebase.

sourcejedi commented:
OK, done.  I would quite like to see at least up to the sixth commit applied.  That one was the motivation for this PR.

I think the new ZCHUNK code got the cleanup correct, it just added yet another duplication of the cleanup lines in prepare_next_transfer().  This PR removes the duplicated cleanup in the `return FALSE` branch.  The `return TRUE` branch is more unique, so I left it where it was (and adjusted it to fit my changes).

The merge history version is at https://github.com/rpm-software-management/librepo/compare/master...sourcejedi:use_cleanup_merge?expand=1 .  GitHub doesn't really understand merge commits though.  `gitk` highlights the conflict resolution lines nicely in black (prefixed with "++").  I've copied & pasted them.

3 conflicts:

```
 +    } else {
 +        // Use supplied filename
 +        int open_flags = O_CREAT|O_TRUNC|O_RDWR;
-         if (target->resume)
++        if (target->resume || target->target->is_zchunk)
 +            open_flags &= ~O_TRUNC;
```

```
@@@ -937,8 -1431,32 +1435,34 @@@
      target->writecb_recieved = 0;
      target->writecb_required_range_written = FALSE;
  
+     #ifdef WITH_ZCHUNK
+     // If file is zchunk, prep it
+     if(target->target->is_zchunk && !check_zck(target, &tmp_err)) {
+         g_set_error(err, LR_DOWNLOADER_ERROR, LRE_ZCK,
+                     "Unable to initialize zchunk file %s: %s",
+                     target->target->path,
+                     tmp_err->message);
 -        curl_easy_cleanup(h);
 -        return FALSE;
++        goto fail;
+     }
+ 
+     // If zchunk is finished, we're done
+     if(target->target->is_zchunk &&
+        target->zck_state == LR_ZCK_DL_FINISHED) {
+         g_debug("%s: Target already fully downloaded: %s", __func__, target->target->path);
+         target->state = LR_DS_FINISHED;
 -        curl_easy_cleanup(h);
++        curl_easy_cleanup(target->handle);
++        target->handle = NULL;
+         g_free(target->headercb_interrupt_reason);
+         target->headercb_interrupt_reason = NULL;
+         fclose(target->f);
+         target->f = NULL;
+         lr_downloadtarget_set_error(target->target, LRE_OK, NULL);
+         return TRUE;
+     }
+     # endif /* WITH_ZCHUNK */
+ 
 +    int fd = fileno(target->f);
 +
      // Allow resume only for files that were originally being
      // downloaded by librepo
      if (target->resume && !has_librepo_xattr(fd)) {
```

```
+         #endif /* WITH_ZCHUNK */
 -            ret = check_finished_trasfer_checksum(fd,
++            ret = check_finished_transfer_checksum(fd,
```

Fourthly, I added a code style change.  This shows up as a separate commit in the merge history version: https://github.com/rpm-software-management/librepo/commit/8955a36fe4b219

Fifthly, the compiler nagged me into adding [another commit on the end](https://github.com/rpm-software-management/librepo/pull/89/commits/ba08d443b5a02420b32341145021f14cd25b1a09). "downloader prepare_next_transfer: remove unused variable".

sourcejedi commented:
F***.  Wait, somehow I *still* didn't fix my mistake that Tojaj pointed, out in check_finished_transfer_checksum().

sourcejedi commented:
Ah, I see what happened.  I had left the "bad" commit, but added a later commit to fix it.  Now I rebased to fix the bad commit.  (The later fix commit was automatically dropped, because it was now empty).  The resulting tree is the same, but the history has been rewritten.

And that's for the best.

Since I had to rebase to master anyway, there was no point leaving the old history intact.  (Again, you can look at the branch `use_cleanup_merge` in my repo.  That still has the "real" history, and uses a merge commit to generate the exact same tree).

Conan-Kudo commented:
@sourcejedi No, this is fine.

But to answer your questions about librepo: it is used by _everything_ involved with package management now, and a number of tools use it independently to fetch repodata for their own purposes. It was created as a separate library to facilitate those use-cases while not making the DNF API itself host those.

Conan-Kudo commented:
@dmach Could you please have someone take a look at this and merge it if it's all gravy?

jrohel commented:
@sourcejedi Thanks for PR. I started review.
But the librepo is not compilable with your PR.
librepo/librepo/downloader.c:1439:57: error: ‘tmp_err’ undeclared (first use in this function)
Problem with ZCHUNK. Please, can you do a small fix?

sourcejedi commented:
Heh. I thought that was odd. This seems to fix it ^.

Thanks for looking at this.

jrohel commented on librepo/downloader.c:
```diff
@@ -1437,27 +1436,30 @@ prepare_next_transfer(LrDownload *dd, gboolean *candidatefound, GError **err)
+            target->f = NULL;
+            lr_downloadtarget_set_error(target->target, LRE_OK, NULL);
+            return TRUE;
+	}
```
Please be consistent. Use 8 spaces instead of tab.

jrohel commented:
@sourcejedi I finished review. Thanks for your work.
I have only one cosmetics request. Please change one line in the last commit. Use 8 spaces instead of tab.

sourcejedi commented:
Ah, I should have configured my editor!  Fixed.

Thanks again.  I will be happy to see this merged.  On the bright side, this pull request has been reviewed by a number of people now, I can hope there are no bugs left in it :-).

jrohel commented:
@sourcejedi Probably you forgot  commit the fix, before you made "force push". I see the old version with tab character.

sourcejedi commented:
Right.  I've committed and pushed now.  Thanks for noticing the whitespace error (twice!).

jrohel commented:
@sourcejedi  Thank you for your contribution and patience. I'm going to merge your PR.
